spring-integration-hazelcast/src/main/java/org/springframework/integration/hazelcast/inbound/HazelcastClusterMonitorMessageProducer.java

```diff
@@ -0,0 +1,254 @@
+ *
+ * @author Eren Avsarogullari
+ * @since 1.0.0
+ */
+public class HazelcastClusterMonitorMessageProducer extends MessageProducerSupport {
+
+	private static final int CLUSTER_MONITOR_TYPE_COUNT = 5;
+
+	private final HazelcastInstance hazelcastInstance;
+
+	private Set<String> monitorTypes = Collections
+			.singleton(ClusterMonitorType.MEMBERSHIP.name());
+
+	private final Map<ClusterMonitorType, String> hazelcastRegisteredListenerIdMap = new ConcurrentHashMap<ClusterMonitorType, String>(
+			CLUSTER_MONITOR_TYPE_COUNT);
+
+	public HazelcastClusterMonitorMessageProducer(
+			final HazelcastInstance hazelcastInstance) {
+		Assert.notNull(hazelcastInstance, "'hazelcastInstance' must not be null");
+		this.hazelcastInstance = hazelcastInstance;
+	}
+
+	public void setMonitorEventTypes(String monitorEventTypes) {
+		HazelcastIntegrationDefinitionValidator.validateEnumType(
+				ClusterMonitorType.class, monitorEventTypes);
+		final Set<String> monitorTypes = StringUtils
+				.commaDelimitedListToSet(monitorEventTypes);
+		Assert.notEmpty(monitorTypes, "'monitorTypes' must have elements");
+		this.monitorTypes = monitorTypes;
+	}
+
+	@Override
+	protected void doStart() {
+		if (this.monitorTypes.contains(ClusterMonitorType.MEMBERSHIP.name())) {
+			final String registrationId = this.hazelcastInstance.getCluster()
+					.addMembershipListener(new HazelcastMembershipListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.MEMBERSHIP,
+					registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.DISTRIBUTED_OBJECT.name())) {
+			final String registrationId = this.hazelcastInstance
+					.addDistributedObjectListener(new HazelcastDistributedObjectListener());
+			this.hazelcastRegisteredListenerIdMap.put(
+					ClusterMonitorType.DISTRIBUTED_OBJECT, registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.MIGRATION.name())) {
+			final String registrationId = this.hazelcastInstance.getPartitionService()
+					.addMigrationListener(new HazelcastMigrationListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.MIGRATION,
+					registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.LIFECYCLE.name())) {
+			final String registrationId = this.hazelcastInstance.getLifecycleService()
+					.addLifecycleListener(new HazelcastLifecycleListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.LIFECYCLE,
+					registrationId);
+		}
+
+		if (this.monitorTypes.contains(ClusterMonitorType.CLIENT.name())) {
+			final String registrationId = this.hazelcastInstance.getClientService()
+					.addClientListener(new HazelcastClientListener());
+			this.hazelcastRegisteredListenerIdMap.put(ClusterMonitorType.CLIENT,
+					registrationId);
+		}
+	}
+
+	@Override
+	protected void doStop() {
+		if (this.hazelcastInstance.getLifecycleService().isRunning()) {
+			if (this.monitorTypes.contains(ClusterMonitorType.MEMBERSHIP.name())) {
```

Comment:
There is no need to do this check. It's just enough make 
```
String id = this.hazelcastRegisteredListenerIdMap.remove(ClusterMonitorType.MEMBERSHIP);
if (id != null) {
    this.hazelcastInstance.getCluster().removeMembershipListener(id);  
}
```

Even if we won't worry about performance, there is a chance that some guy may change `monitorTypes` at runtime.

From other side it looks better to use `bit mask` here and have only one `int` field instead of that heavy `Map`.